File: backend/task-service/src/main/java/org/example/taskservice/model/Task.java
package org.example.taskservice.model;


import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Version;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import org.example.events.enums.Department;
import org.example.events.enums.TaskPriority;
import org.example.events.enums.TaskStatus;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.*;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@Document(collection = "tasks")
@CompoundIndexes({
        @CompoundIndex(name = "status_department_idx", def = "{'status': 1, 'department': 1}", partialFilter = "{status : 'AVAILABLE'}"),
        @CompoundIndex(name = "assignee_status_idx", def = "{'assigneeIds': 1, 'status': 1}")
        }
)
@Data
public class Task {

    @Id
    private String id;

    @NotNull
    private String title;

    private String description;

    @NotNull
    @Indexed(unique=false)
    @Enumerated(EnumType.STRING)
    private TaskStatus status;

    @Enumerated(EnumType.STRING)
    private TaskPriority priority;

    @Indexed
    private Set<String> assigneeIds;

    @Indexed
    private String creatorId;

    @Indexed
    @Enumerated(EnumType.STRING)
    private Department department;

    @NotNull
    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Indexed(direction = IndexDirection.DESCENDING)
    private Instant dueDate;

    private List<String> tags;

    private List<Comment> comments = new ArrayList<>();
}

File: frontend/taskflow-orchestrator/src/types/Task.ts
// types/Task.ts
// @ts-ignore
export enum TaskStatus {
  AVAILABLE = 'AVAILABLE',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  // Добавь другие статусы, если они есть
}

// @ts-ignore
export enum TaskPriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  // Добавь другие приоритеты, если они есть
}

// @ts-ignore
export enum Department {
  DEVELOPMENT = 'DEVELOPMENT',
  SALES = 'SALES',
  HR = 'HR',
  // Добавь другие отделы, если они есть
}

export interface Comment {
  id?: string; // Опционально, если бэкенд генерирует
  text: string;
  createdBy?: string;
  createdAt?: Date;
  // Добавь другие поля комментария, если они есть
}

export interface Task {
  id: string;
  title: string;
  description?: string;
  status: TaskStatus;
  priority?: TaskPriority;
  assigneeIds: Set<string>; // Используем Set для соответствия бэкенду
  creatorId: string;
  department: Department;
  createdAt: Date;
  updatedAt?: Date;
  dueDate?: Date;
  tags?: string[];
  comments?: Comment[];
}